modules/ximgproc/src/scansegment.cpp

```diff
@@ -0,0 +1,926 @@
+
+			cv::Mat src, labelsMat;	// mats
+
+			struct WSNode
+			{
+				int next;
+				int mask_ofs;
+				int img_ofs;
+			};
+
+			// Queue for WSNodes
+			struct WSQueue
+			{
+				WSQueue() { first = last = 0; }
+				int first, last;
+			};
+
+			class PP1 : public cv::ParallelLoopBody
+			{
+			public:
+				PP1(ScanSegmentImpl* const scanSegment)
+					: ss(scanSegment) {}
+				virtual ~PP1() {}
+
+				void operator()(const cv::Range& range) const CV_OVERRIDE
+				{
+					for (int v = range.start; v < range.end; v++)
+					{
+						ss->OP1(v);
+					}
+				}
+			private:
+				ScanSegmentImpl* const ss;
+			};
+
+			class PP2 : public cv::ParallelLoopBody
+			{
+			public:
+				PP2(ScanSegmentImpl* const scanSegment, std::vector<std::pair<int, int>>* const countVec)
+					: ss(scanSegment), ctv(countVec) {}
+				virtual ~PP2() {}
+
+				void operator()(const cv::Range& range) const CV_OVERRIDE
+				{
+					for (int v = range.start; v < range.end; v++)
+					{
+						ss->OP2((*ctv)[v]);
+					}
+				}
+			private:
+				ScanSegmentImpl* const ss;
+				std::vector<std::pair<int, int>>* ctv;
+			};
+
+			class PP3 : public cv::ParallelLoopBody
+			{
+			public:
+				PP3(ScanSegmentImpl* const scanSegment)
+					: ss(scanSegment) {}
+				virtual ~PP3() {}
+
+				void operator()(const cv::Range& range) const CV_OVERRIDE
+				{
+					for (int v = range.start; v < range.end; v++)
+					{
+						ss->OP3(v);
+					}
+				}
+			private:
+				ScanSegmentImpl* const ss;
+			};
+
+			class PP4 : public cv::ParallelLoopBody
+			{
+			public:
+				PP4(ScanSegmentImpl* const scanSegment, std::vector<std::pair<int, int>>* const countVec)
+					: ss(scanSegment), ctv(countVec) {}
+				virtual ~PP4() {}
+
+				void operator()(const cv::Range& range) const CV_OVERRIDE
+				{
+					for (int v = range.start; v < range.end; v++)
+					{
+						ss->OP4((*ctv)[v]);
+					}
+				}
+			private:
+				ScanSegmentImpl* const ss;
+				std::vector<std::pair<int, int>>* ctv;
+			};
+
+			void OP1(int v);
+			void OP2(std::pair<int, int> const& p);
+			void OP3(int v);
+			void OP4(std::pair<int, int> const& p);
+			void expandCluster(int* offsetBuffer, const cv::Point& point);
+			void calculateCluster(int* offsetBuffer, int* offsetEnd, int pointIndex, int currentClusterID);
+			static int allocWSNodes(std::vector<WSNode>& storage);
+			static void watershedEx(const cv::Mat& src, cv::Mat& dst);
+		};
+
+		CV_EXPORTS Ptr<ScanSegment> createScanSegment(int image_width, int image_height, int num_superpixels, int threads, bool merge_small)
+		{
+			return makePtr<ScanSegmentImpl>(image_width, image_height, num_superpixels, threads, merge_small);
+		}
+
+		ScanSegmentImpl::ScanSegmentImpl(int image_width, int image_height, int num_superpixels, int threads, bool merge_small)
+		{
+			// set the number of process threads
+			processthreads = std::thread::hardware_concurrency();
+			if (threads > 0) {
+				processthreads = MIN(processthreads, threads);
```

Comment:
As different "threads" values provide different results it is better to avoid changing this parameter depending on runtime environment.

It makes sense to rename this parameter to "slices" or similar.

Reply:
parameter changed to slices. default slices set to 8. However, setting to -1 will use the maximum number of processing threads. If the number of cpu cores from std::thread::hardware_concurrency() is less than this setting, it will use the smaller of the two to set the number of processing threads.

Reply:
I had a think about this and set the number of processing threads to be equal to the default slices. This should ensure consistent results. The user is still able to change this value to another number or even -1, and this will offer the option of speeding the algorithm if consistent results is not a priority.